pkg/multimap/composite.go

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) 2019-Present Pivotal Software, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package multimap
+
+import "fmt"
+
+// Composite is an interface to a consistent collection of mappings from string to string. The mappings are consistent
+// in the sense that no two mappings map a given value to distinct results.
+type Composite interface {
+	// Add updates the composite mapping by adding a mapping with the given name. If a mapping with the given name
+	// already exists, it is replaced providing it is consistent with all other mappings. If the named mapping is not
+	// consistent with all other mappings, the mapping is removed from the composite mapping and error is returned.
+	Add(name string, mapping map[string]string) error
+
+	// Delete updates the composite mapping by removing a mapping with the given name. If there is no mapping with the
+	// given name, an error is returned.
+	Delete(name string) error
+
+	// Map applies the composite mapping to the given value and returns the mapped value. If the given value is not
+	// in the domain of the composite mapping, the given value is returned. In other words, the default mapping is
+	// the identity function.
+	Map(string) string
+}
+
+type errCh chan error
+
+type addOp struct {
+	name    string
+	mapping map[string]string
+	errCh   errCh
+}
+
+type deleteOp struct {
+	name  string
+	errCh errCh
+}
+
+type mapOp struct {
+	value    string
+	resultCh chan string
+}
+
+type composite struct {
+	// a consistent collection of mappings
+	mappings map[string]map[string]string
+
+	// the composition of all the mappings
+	composite map[string]string
+
+	addCh    chan *addOp
+	deleteCh chan *deleteOp
+	mapCh    chan *mapOp
+	stopCh   <-chan struct{}
+}
+
+func New(stopCh <-chan struct{}) Composite {
+	c := &composite{
```

Comment:
I doubt this needs to be so complex. The reconciler can be run single threaded (if not by default) so there should not be contention on mutations.

Reply:
That wasn't my concern. The webhook is a HTTP server and so it runs on distinct goroutines to the controller.